streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -299,7 +318,17 @@ private void restoreState(final StateRestoreCallback stateRestoreCallback,
                             restoreRecords.add(recordConverter.convert(record));
                         }
                     }
-                    offset = globalConsumer.position(topicPartition);
+                    try {
+                        offset = globalConsumer.position(topicPartition);
+                    } catch (final TimeoutException error) {
+                        // the `globalConsumer.position()` call should never block, because we know that we did
+                        // a successful `position()` call above for the requested partition and thus the consumer
+                        // should have a valid local position that it can return immediately
```

Comment:
I think this makes sense, but it seems to make some assumptions about the internal implementation of the consumer that doesn't seem necessary here. Is it important to assert that the consumer can't possibly get a timeout exception here?

Reply:
I guess strictly speaking, it's not necessary, but I personally prefer this strict style as it surfaces bugs quicker. If we write "robust" code, it could hide a bug. -- It's just a personal preference and we have similar code elsewhere. If you feel very strong about it, I am also ok to change it.

Reply:
Thanks; yes, this is a very good habit. In this specific case, I wasn't sure what the "bug" would be, though. Just that the implementation of the consumer does something different, but equally correct, from our current understanding of how it works?

Reply:
The bug would be in the consumer: it forgets its position and needs to fetch it from the broker again (of course, we would only expose this bug, if there is also a broker/network issue and the request actually times out... what makes it a rather weak "check").

Reply:
We don't need to call `retryUntilSuccessOrThrowOnTaskTimeout` now, so we actually get cleaner code :)

Reply:
It seems like what we're doing here is syncing our `offset` variable with the internal position cursor of the client. Although @mjsax is claiming the client never needs to update its position from the consumer group, the position would advance as a consequence of polling records earlier in this loop. Right?

Reply:
Hey @mjsax , it sounds like you have an opinion about what would be unexpected internal behavior of the client, based on your knowledge of how it is currently implemented. There doesn't seem to be anything in the contract of the client that says we should rely on this implementation detail, so it seems inappropriate to build in a fatal exception for this case.

Reply:
What John says: We need to track the restore progress to eventually break the outer while loop. We don't use `offset` to `seek()`, so it's not about updating the consumer's position but knowing when we are done.

Reply:
I guess with the new `retryUntilSuccessOrThrowOnTaskTimeout` the code is still clean is we just use it here.